lib/backend/k8s/resources/node.go

```diff
@@ -282,6 +300,113 @@ func mergeCalicoNodeIntoK8sNode(calicoNode *apiv3.Node, k8sNode *kapiv1.Node) (*
 	return k8sNode, nil
 }
 
+// mergeCalicoAndK8sLabels merges the Kubernetes labels (from k8sNode.Labels) with those that are already present in
+// calicoNode (which were loaded from our annotation).  Kubernetes labels take precedence.  To make the operation
+// reversible (so that we can support write back of a Calico node that was read from Kubernetes), we also store the
+// complete set of Kubernetes labels in an annotation along with any Calico labels that were shadowed by Kubernetes
+// labels.
+func mergeCalicoAndK8sLabels(calicoNode *apiv3.Node, k8sNode *kapiv1.Node) error {
+	// Now, copy the Kubernetes Node labels over.  Note: this may overwrite Calico labels of the same name, but that's
+	// consistent with the kube-controllers behavior.
+	var shadowedLabels map[string]string
+	for k, v := range k8sNode.Labels {
+		if calicoNode.Labels == nil {
+			calicoNode.Labels = map[string]string{}
+		}
+		if calValue, ok := calicoNode.Labels[k]; ok {
+			if shadowedLabels == nil {
+				shadowedLabels = map[string]string{}
+			}
+			shadowedLabels[k] = calValue
+		}
+		calicoNode.Labels[k] = v
+	}
+
+	// For consistency with kube-controllers, and so we can correctly round-trip labels, we stash the kubernetes labels
+	// in an annotation.
+	if calicoNode.Annotations == nil {
+		calicoNode.Annotations = map[string]string{}
+	}
+	bytes, err := json.Marshal(k8sNode.Labels)
+	if err != nil {
+		log.WithError(err).Errorf("Error marshalling node labels")
+		return err
+	}
+	calicoNode.Annotations[nodeK8sLabelAnnotation] = string(bytes)
+
+	// To be able to round trip Calico labels, we stash any labels that we've shadowed in an annotation.
+	if shadowedLabels != nil {
+		bytes, err := json.Marshal(shadowedLabels)
+		if err != nil {
+			log.WithError(err).Errorf("Error marshalling node labels")
+			return err
+		}
+		calicoNode.Annotations[nodeShadowedLabelAnnotation] = string(bytes)
+	}
+	return nil
+}
+
+// restoreCalicoLabels tries to undo the transformation done by mergeCalicoLabels.  It returns a copy of the node with
+// Kubernetes labels removed from the Labels field and restores any shadowed Calico labels.  It also filters the
+// bookkeeping annotations out of the copy.
+func restoreCalicoLabels(calicoNode *apiv3.Node) (*apiv3.Node, error) {
+	calicoNode = calicoNode.DeepCopy()
```

Comment:
That would do a shallow copy of the Node, not a deep copy.  A shallow copy copies all the fields of the struct by value.  For simple types such as ints that works fine but for reference types (such as maps) it would only copy the reference, so we'd end up sharing the labels between the original and copied object.

In an earlier iteration of the code, I did use a shallow copy and then I manually copied the labels when needed; however, it added complexity and I found that I wrote more bugs than Id like in such a short piece of code so I switched to a  deep copy operation.  It probably does perform slightly worse but it's a lot simpler.